cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -35,6 +34,90 @@ namespace lists {
 namespace detail {
 namespace {
 using offset_type = lists_column_view::offset_type;
+
+/**
+ * @brief Performs an equality comparison between two rows in a single column
+ *
+ * If the row elements are not of floating point types, this functor will return the
+ * same comparison result as `cudf::element_equality_comparator`.
+ * For floating point types, elements that hold NaN value will be considered as unequal.
+ *
+ * @tparam has_nulls Indicates the potential for null values in the column
+ */
+template <bool has_nulls = true>
+class element_comparator_fn {
+ public:
+  /**
+   * @brief Construct type-dispatched function object for comparing equality
+   * between two rows
+   *
+   * @param d_view The column containing the rows to compare
+   * @param nulls_are_equal Indicates if two null rows are treated as having the same value
+   */
+  __host__ __device__ element_comparator_fn(column_device_view d_view, bool nulls_are_equal)
+    : d_view{d_view}, nulls_are_equal{nulls_are_equal}
+  {
+  }
+
+  template <class T, std::enable_if_t<cudf::is_equality_comparable<T, T>()>* = nullptr>
+  __device__ bool operator()(size_type i, size_type j) const noexcept
+  {
+    if (has_nulls) {
+      bool const nullable = d_view.nullable();
+      bool const lhs_is_null{nullable and d_view.is_null(i)};
+      bool const rhs_is_null{nullable and d_view.is_null(j)};
+      if (lhs_is_null and rhs_is_null) {
+        return nulls_are_equal;
+      } else if (lhs_is_null != rhs_is_null) {
+        return false;
+      }
+    }
+
+    // For floating point types, if both element(i) and element(j) are NaNs then this comparison
+    // will return `false`. This is the desired behavior for `drop_list_duplicates`.
+    return d_view.element<T>(i) == d_view.element<T>(j);
+  }
+
+  template <typename Element,
+            std::enable_if_t<not cudf::is_equality_comparable<Element, Element>()>* = nullptr>
+  __device__ bool operator()(size_type, size_type) const
+  {
+    release_assert(false && "Attempted to compare elements of uncomparable types.");
+    return false;
+  }
+
+ private:
+  column_device_view d_view;
+  bool nulls_are_equal;
+};
+
+/**
+ * @brief Perform an equality comparison between two rows in a single column
+ *
+ * @tparam has_nulls Indicates the potential for null values in the column
+ */
+template <bool has_nulls>
+class element_comparator {
+ public:
+  element_comparator(column_device_view _d_view, null_equality _nulls_equal)
+    : d_view{_d_view}, nulls_equal{_nulls_equal}
+  {
+  }
+
+  __device__ bool operator()(size_type i, size_type j) const noexcept
+  {
+    return cudf::type_dispatcher(
```

Comment:
If it's only a single column, then the type dispatch can happen once in the host code instead of doing it for ever comparison in device code.

Reply:
+1

Reply:
Good point. Thanks, Jake. I have updated my code accordingly.